src/oni2-keyboard-layout/stubs/keyboard-layout-mac.c

```diff
@@ -0,0 +1,240 @@
+#ifdef __APPLE__
+
+#import <Cocoa/Cocoa.h>
+#import <Carbon/Carbon.h>
+
+#include <caml/alloc.h>
+#include <caml/callback.h>
+#include <caml/memory.h>
+#include <caml/mlvalues.h>
+#include <caml/threads.h>
+
+#include "keyboard-layout.h"
+
+typedef struct KeycodeMapEntry {
+  uint16_t virtualKeyCode;
+  const char *dom3Code;
+} KeycodeMapEntry;
+
+#define USB_KEYMAP_DECLARATION static const KeycodeMapEntry keyCodeMap[] =
+#define USB_KEYMAP(usb, evdev, xkb, win, mac, code, id) {mac, code}
+
+#define UNWRAP_REF(ref) (Field(ref, 0))
+
+#include "keycode_converter_data.h"
+
+NSString *oni2_priv_GetCurrentKeyboardLayout() {
+  TISInputSourceRef source = TISCopyCurrentKeyboardInputSource();
+  CFStringRef sourceID = (CFStringRef)TISGetInputSourceProperty(source, kTISPropertyInputSourceID);
+
+  return (NSString *)sourceID;
+}
+
+NSString *oni2_priv_GetCurrentKeyboardLanguage() {
+  TISInputSourceRef source = TISCopyCurrentKeyboardInputSource();
+  NSArray *langs = (NSArray *)TISGetInputSourceProperty(source, kTISPropertyInputSourceLanguages);
+  NSString *lang = (NSString *)[langs objectAtIndex:0];
+
+  return lang;
+}
+
+static void notificationHandler(
+  CFNotificationCenterRef center, 
+  void *manager,
+  CFStringRef name, 
+  const void *object, 
+  CFDictionaryRef userInfo
+) {
+  CAMLparam0();
+  CAMLlocal2(vLayout, vLanguage);
+  // Maintain a reference to the callbacks from the OCcaml side
+  // We let OCaml manage the list since they have better abstractions than we do
+  static const value *callbackListRef = NULL;
+  // If we haven't gotten the reference yet, get it.
+  if (callbackListRef == NULL) {
+    callbackListRef = caml_named_value("oni2_KeyboardLayoutCallbackListRef");
+    // If we didn't get anything back from the OCaml runtime, return early.
+    if (callbackListRef == NULL) {
+      NSLog(@"Unable to acquire callback list reference!");
+      return;
+    }
```

Comment:
This `return` will actually cause problems, since we would need a `CAMLreturn` to work properly with the garbage collector. But if we streamline the callback strategy as mentioned in an earlier comment - we shouldn't even need this logic here.

Reply:
I might be mistaken, but I think I still need to check if the pointer returned by `caml_named_value` is null, or else I would get a segfault when I de-reference it in `caml_callbackN`. Please let me know if I'm wrong!